### PR TITLE
Fix conflicting key in init kwargs in PreTrainedTokenizerBase

### LINKED ISSUE #30796: Serialization error when tokenizer_config key matches function name in PreTrainedTokenizerBase
https://github.com/huggingface/transformers/blob/37bba2a32d2742a10216ffd925bb8f145a732ce1/src/transformers/tokenization_utils_base.py#L2449-L2451

When one of the keys in `self.init_kwargs` matches the name of a function in PreTrainedTokenizerBase (e.g., `add_special_tokens`), this for loops replaces the value for that key in `tokenizer_config` with the function object, which is not serializable, thus causing an error during `save_pretrained`.

To solve this issue, one option is to add an assert in the `__init__` function that throws an error if one of the keys matches an existing attribute/function on the `PreTrainedTokenizerBase`:
https://github.com/huggingface/transformers/blob/37bba2a32d2742a10216ffd925bb8f145a732ce1/src/transformers/tokenization_utils_base.py#L1569

This error was also raised in the Stack Overflow issue below:
https://stackoverflow.com/questions/78062739/huggingface-transformers-error-when-saving-model-typeerror-object-of-type-meth

### PR DESCRIPTION
**Pull request by OmarManzoor**
# What does this PR do?
Checks if there is any key that matches a method name in PreTrainedTokenizerBase and raises an error if so to prevent errors that occur during serialization.

Fixes #30796

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?

## Who can review?
@ArthurZucker @amyeroberts

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
LGTM, tho I think if someone tries to save `add_special_tokens` it's something we would want to save to re-use when encoding.

Might be the opposite of what I said, but saving it can be convenient WDYT?

**Comment by OmarManzoor** [comment]
@ArthurZucker I added the check for a conflicting method in the save_pretrained method. Is this suitable?

**Review by ArthurZucker — APPROVED** [review]
I think this is totally mergable! 
Let's think a tad bit about `add_special_tokens`, I think we could use a different name like `_add_special_tokens` which will be set when use `tokenizer.add_special_tokens` with a setter. Might be possible? 
otherwise we will go with adding a new argument for this and deprecated `add_special_tokens`

**Comment by OmarManzoor** [comment]
We can set a field `_add_special_tokens` which makes sense, but in that case where this field is used has to be changed?

**Review by ArthurZucker — COMMENTED** [review]
Yep agree with you would be a bit annoying. 
Let's check if callable, then raise an error. Regarding add_special_tokens, we can use self._add_special_tokens in encode for example which today does not use it (no places uses it since it is not saved as of today)

**Review by ArthurZucker — APPROVED** [review]
thanks for the fix

**Comment by ArthurZucker** [comment]
Could you re-commit something the CI does not seem to start

**Comment by OmarManzoor** [comment]
@ArthurZucker I am not sure about the failing tests.

**Comment by ArthurZucker** [comment]
Unrelated no worries

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -1569,6 +1569,13 @@ class PreTrainedTokenizerBase(SpecialTokensMixin, PushToHubMixin):
     def __init__(self, **kwargs):
         # inputs and kwargs for saving and re-loading (see ``from_pretrained`` and ``save_pretrained``)
         self.init_inputs = ()
+        if "add_special_tokens" in kwargs:
+            kwargs["_add_special_tokens"] = kwargs.pop("add_special_tokens")
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:1573]: why are we saving this? It's never used btw no?

Comment (OmarManzoor) [reply src/transformers/tokenization_utils_base.py:1573]: I think you mentioned we might be able to use it.
https://github.com/huggingface/transformers/pull/31233#pullrequestreview-2125432908

Comment (ArthurZucker) [reply src/transformers/tokenization_utils_base.py:1573]: Yeah but if we don't then let's not save it!

**Review by ArthurZucker — APPROVED** [review]
Yeah LGTM, IMO should not slow down init as we check `hasattr`, which should be based on name of the kwargs -> O(1)

**Pull request merged**